lambda/matching.ml

```diff
@@ -2819,48 +2878,109 @@ let combine_constructor loc arg pat_env cstr partial ctx def
-                        call_switcher loc fail_opt arg 0 (n - 1) consts,
-                        act )
+                let single_nonconst_act =
+                  single_action
+                    cases.any_nonconst cases.nonconsts vd.vd_nonconsts in
+                let single_const_act =
+                  single_action
+                    cases.any_const cases.consts vd.vd_consts in
+                let test_isint const_act nonconst_act =
+                  Lifthenelse
+                    ( Lprim (Pisint, [ arg ], loc), const_act, nonconst_act )
+                in
+                (* We check specifically for single_nonconst_act, not for
+                   single_const_act: for constant constructors we can generate
+                   better code than a switch in some cases, but for tests on
+                   non-constant constructors we prefer to always emit a switch,
+                   as bytecode implements this sophisticated instruction. *)
+                match single_nonconst_act with
+                | Some nonconst_act ->
+                   (* Note: we already checked that not all actions are
+                      identical, so the constant-constructor actions cannot
+                      be empty. *)
+                   let int_actions =
+                     match single_const_act with
+                     | Some const_act -> const_act
+                     | None ->
+                        call_switcher loc fail_opt arg 0 (n - 1) cases.consts
+                   in test_isint int_actions nonconst_act
                 | None ->
-                    (* Emit a switch, as bytecode implements this sophisticated
-                      instruction *)
+                    let fail_opt =
+                      match fail_opt, cases.any_const with
+                      | None, None -> None
+                      | Some a, None
+                      | None, Some a ->
+                          Some a
+                      | Some fail_act, Some a ->
+                          Some (test_isint a fail_act)
+                    in
+                    let numconsts, numblocks =
+                      let bounds = match vd.vd_max_values with
+                        | Some bounds -> bounds
+                        | None ->
+                          (* The values are not filled e.g in case a type is
+                             imported from a foreign cmi interface; we need to
+                             recompute them here *)
+                          if vd.vd_unboxed = 0 then
+                            Max vd.vd_consts, Max vd.vd_nonconsts
+                          else
+                            Typedecl_unboxed.Head_shape.max_val_of_cstr_descr
+                              pat_env cstr
+                      in match bounds with
+                        | Max imm, Max tag -> imm, tag
+                        | Unbounded, _ | _, Unbounded ->
+                            invalid_arg "Matching.combine_constructors"
```

Comment:
We know that we have `single_nonconst_act = None` here, so unless I am mistaken the `_, Unbounded` case can never happen (it could be turn into an `assert false` with a comment explaining why it is unreachable.) On the other hand, why is the `Unbounded, _` case impossible? I would think that it is possible with something like:

```ocaml
type t = Int of int [@unboxed] | Cstr1 of unit | Cstr2 of unit`

let test = function
  | Int n -> n
  | Cstr1 -> 1
  | Cstr2 -> 2
```

If we have `Shape_any` on immediate values, we should generate a switch with `sw_numconsts = 0`, which should work well with our definition of `fail_opt` above (it will in fact go to the `any_const` action on immediates). This is not what the code below is doing, so I would guess we have a bug? (But then I have not tested anything.)